File: cpp_10-99/31.cc
// C++
// leetcode 31
// https://leetcode-cn.com/problems/next-permutation/
// 模拟，偷三叶的方法

class Solution {
public:
    void reverse(vector<int>& nums, int l, int r) {
        while(l < r) {
            swap( nums[l], nums[r]);
            ++l;
            --r;
        }
    }

    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int p = n - 1;
        while(p >= 1 && nums[p - 1] >= nums[p])
            --p;
        if(p == 0) {
            reverse(nums, 0, n - 1);
            return;
        } else {
            int q = p;
            while(q + 1 < n && nums[q + 1] > nums[p - 1])
                ++q;
            swap(nums[p - 1], nums[q]);
            reverse(nums, p, n - 1);
        }
    }
};

File: Go/Go_10-99/31.go
// Go
// leetcode 31
// https://leetcode.cn/problems/next-permutation/

func nextPermutation(nums []int) {
	n := len(nums)
	if n <= 1 {
		return
	}

	i, j, k := n-2, n-1, n-1

	// find nums[i] < nums[j]
	for i >= 0 && nums[i] >= nums[j] {
		i--
		j--
	}

	if i >= 0 { // 不是最后一个排列
		// find nums[i] < nums[k]
		for nums[i] >= nums[k] {
			k--
		}

		// swap
		nums[i], nums[k] = nums[k], nums[i]
	}

	for i, j = j, n-1; i < j; i, j = i+1, j-1 {
		nums[i], nums[j] = nums[j], nums[i]
	}
}